Add ability to create shared folders

Naomi-alt opened the pull request:
So I wanted to give a try at implemented a feature that I (and I think many others) have wanted for some time now. (See #38,  #1027  )

It's my first time working with QT and this code base, so any feedback is welcome!

Here's my plan:
- [x] Add 3 checkboxes in the instance settings (under Settings->Miscellaneous->Shared folders) to enable/disable the shared directory for screenshots/saves/resource packs.
  - If the original directory is not empty, give an error.
  - If the original directory does not exist or is empty, create a symbolic link to a hardcoded directory. (screenshots, saves or resourcepacks)
Maybe also add a tooltip/warning that when disabling the shared directory, the items will stay inside the shared directory and not be moved to the instance specific directory.
- [x] Add a warning to the Screenshots, Worlds and Resource packs settings when the shared directory is enabled that the items you're looking at are shared across instances.
- [x] Allow customizable shared directory.
  - When checkbox is checked, an input field can be enabled which defaults to the default shared directory, but can be 
customized to an other location.
  - [x] Test: When browsing, my file picker stops at the first encountered hidden directory. So in my case, it opens at `/home/naomi` as it can't find the `.local/`. My filepicker has an option to enable hidden files, but that isn't saved. This is probably a bug with my setup, and not necessarily qt?
- [x] Allow non-empty directories to be turned into shared directories
  - If a directory is not empty, copy the contents to the shared directory.
  - If the contents already exists, ask the user what to do. (Similar to Windows File Explorer? Is there already a UI/component for this, or should I create one?)
  - If everything could be successfully moved, delete instance directory and create symlink.

Nice to haves (most likely a future pull-request):
- Add shared default checkboxes.
  - Set the shared folder for each instance.
- Allow setting other directories to be shared, for example, schematics.
  - Basically a convenient way for non-technical people to create symlinks?
  - A scroll-list with your created symlinks under Settings->Miscellaneous->shared folders
  - The merging of existing files could be handled the same way the other directories are handled.
- Shared server list
  - This is most likely a whole other feature, since you'd have to merge the servers.dat but it could be fun challenge to try and tackle some day.

Some questions I still have:
- I wanted to separate the logic of applying the settings from the ui component, hence the `MinecraftInstance::applySettings` routine. But I'm not sure if the current implementation is correct/desired.
  Ideally you'd want the settings to be applied immediately, so you can get a relevant error as soon as possible. But I was told that would require a decent overhaul, since you'd have to change when all settings are applied to keep everything consistent. So what is the correct solution for the mean time?
- How much should I mention the existence symbolic links to the user? I doubt the average user knows what they are and how they work.
- I probably messed up the UI. Any tips on working with it?

Naomi-alt committed:
Create initial global folders feature

Signed-off-by: Naomi <103967@gmail.com>

Naomi-alt committed:
Fix formatting

Signed-off-by: Naomi <103967@gmail.com>

Naomi-alt committed:
Make global folder path customizable

Signed-off-by: Naomi <103967@gmail.com>

Naomi-alt committed:
Add file conflict dialog

Signed-off-by: Naomi <103967@gmail.com>

Ryex commented on launcher/minecraft/UpdateGlobalDirectoriesTask.cpp:
```diff
@@ -0,0 +1,182 @@
+    if (!sourceInfo.exists())
+        return false;
+
+    if (recursive) {
+        // Recursive doesn't make sense if the source isn't a directory.
+        if (!sourceInfo.isDir())
+            return false;
+
+        QDirIterator sourceIt(source, QDir::Filter::Files | QDir::Filter::Dirs | QDir::Filter::Hidden | QDir::Filter::NoDotAndDotDot);
+
+        while (sourceIt.hasNext()) {
+            if (!interactiveMove(sourceIt.next(), FS::PathCombine(destination, sourceIt.fileName()), false))
+                return false;
+        }
+
+        return true;
+    }
```
```suggestion
    if (recursive && sourceInfo.isDir()) {
        // Recursive doesn't make sense if the source isn't a directory.

        QDirIterator sourceIt(source, QDir::Filter::Files | QDir::Filter::Dirs | QDir::Filter::Hidden | QDir::Filter::NoDotAndDotDot);

        while (sourceIt.hasNext()) {
            if (!interactiveMove(sourceIt.next(), FS::PathCombine(destination, sourceIt.fileName()), false))
                return false;
        }

        return true;
    }
```


IMO it makes much more sense to just ignore the recursion param if the source isn't a directory

Ryex commented on launcher/minecraft/UpdateGlobalDirectoriesTask.cpp:
```diff
@@ -0,0 +1,182 @@
+
+        FS::create_link folderLink(m_source, m_destination);
+        folderLink.linkRecursively(false);
+        folderLink();  // TODO: Error check
+
```
don't forget your todos :)

Naomi-alt replied:
Agreed :)

Ryex commented on launcher/ui/dialogs/FileConflictDialog.cpp:
```diff
@@ -0,0 +1,85 @@
+    }
+
+    // Setup info
+    QFileInfo sourceInfo(source);
+    ui->sourceInfoLabel->setText(tr("<b>Name:</b> %1<br/><b>Size:</b> %2<br/><b>Last modified:</b> %3")
+                                     .arg(source, sourceInfo.isDir() ? "-" : locale.formattedDataSize(sourceInfo.size()),
+                                          sourceInfo.lastModified().toString(locale.dateTimeFormat())));
+
+    QFileInfo destinationInfo(destination);
+    ui->destinationInfoLabel->setText(tr("<b>Name:</b> %1<br/><b>Size:</b> %2<br/><b>Last modified:</b> %3")
+                                          .arg(destination,
+                                               destinationInfo.isDir() ? "-" : locale.formattedDataSize(destinationInfo.size()),
+                                               destinationInfo.lastModified().toString(locale.dateTimeFormat())));
```
it might be nice to show a count of directory contents when the source or dest is a directory. not a blocker though

Naomi-alt replied:
Good idea!

Ryex reviewed (changes requested):
I noticed that you marked it as being ready so I did a bit of code review. havn't tested yet.

You using a per instance setting (which would be desired behavior), but your not providing a global default setting.

I think that this should be changed so that there is a global default settings for the folders with an instance setting that overrides the global defaults to use the instance's settings.
Look at the java override setting for an example of what I mean.

Naomi-alt commented:
> I noticed that you marked it as being ready so I did a bit of code review. havn't tested yet.
> 
> You using a per instance setting (which would be desired behavior), but your not providing a global default setting.
> 
> I think that this should be changed so that there is a global default settings for the folders with an instance setting that overrides the global defaults to use the instance's settings. Look at the java override setting for an example of what I mean.

Yeah I was planning on adding that eventually, but I haven't exactly figured out how I want that to look.
As in, if you enable the default checkbox for the global saves folder, and you have a lot of instances with "New World", you might get a lot of FileConflictDialogs thrown at you.
If you think that's not a problem, I'll go ahead and implement it, else I've got to come up with a good solution.

Ryex commented:
> > I noticed that you marked it as being ready so I did a bit of code review. havn't tested yet.
> > You using a per instance setting (which would be desired behavior), but your not providing a global default setting.
> > I think that this should be changed so that there is a global default settings for the folders with an instance setting that overrides the global defaults to use the instance's settings. Look at the java override setting for an example of what I mean.
> 
> Yeah I was planning on adding that eventually, but I haven't exactly figured out how I want that to look. As in, if you enable the default checkbox for the global saves folder, and you have a lot of instances with "New World", you might get a lot of FileConflictDialogs thrown at you. If you think that's not a problem, I'll go ahead and implement it, else I've got to come up with a good solution.

Treat an instance that doesn't have the override setting set (as in the instance was created before the setting was introduced) as having the override setting on and all global folders disabled

Naomi-alt committed:
Review feedback

Signed-off-by: Naomi <103967@gmail.com>

Naomi-alt committed:
Implement global overridable global-folders-settings

Signed-off-by: Naomi <103967@gmail.com>

Ryex commented:
This looks good now, I dont have the time now but it needs some testing before final approval

Trial97 commented on launcher/BaseInstance.cpp:
```diff
@@ -105,6 +106,17 @@ BaseInstance::BaseInstance(SettingsObjectPtr globalSettings, SettingsObjectPtr s
+    // Global folders
+    auto globalScreenshotsSetting = m_settings->registerSetting("OverrideGlobalScreenshots", false);
+    m_settings->registerOverride(globalSettings->getSetting("UseGlobalScreenshotsFolder"), globalScreenshotsSetting);
+    m_settings->registerOverride(globalSettings->getSetting("GlobalScreenshotsPath"), globalScreenshotsSetting);
```
Do we need the path to the global stuff to be controlled per instance?

Naomi-alt replied:
Are you talking about the fact that each instance can have their own path?
If so, than that is by design. The default global folder is just that, a default. If the user so desires they can have multiple global folders.
For example: they may want to keep their vanilla and modded global folders separate, but still have a shared folder within each group.

Trial97 replied:
Ok then on UI level, I want a change: from a global folder to a shared folder at least for the instance settings.
As global should be unique and in this case is not.

Naomi-alt replied:
Done. Global folders are now shared folders, which is indeed more clear of what it does.

Trial97 commented on launcher/ui/pages/instance/WorldListPage.cpp:
```diff
@@ -120,6 +120,9 @@ void WorldListPage::openedImpl()
     ui->toolBar->setVisibilityState(m_wide_bar_setting->get().toByteArray());
+
+    // Enable the symbolic link warning when the saves folder is a symbolic link
+    ui->globalSavesFolderWarninglabel->setVisible(FS::isSymLink(FS::PathCombine(m_inst->gameRoot(), "saves")));
```
This doesn't guarantee that it points towards a global folder.
You should update the label text to reflect this or make this check more strict(check if the link points towards the global folder)

Naomi-alt replied:
Updated the warning text

Trial97 commented on launcher/ui/pages/instance/ScreenshotsPage.cpp:
```diff
@@ -560,6 +560,9 @@ void ScreenshotsPage::openedImpl()
     ui->toolBar->setVisibilityState(m_wide_bar_setting->get().toByteArray());
+
+    // Enable the symbolic link warning when the screenshots folder is a symbolic link
+    ui->globalScreenshotsFolderWarninglabel->setVisible(FS::isSymLink(m_folder));
```
same as for worlds

Trial97 commented on launcher/ui/pages/instance/ExternalResourcesPage.cpp:
```diff
@@ -138,6 +139,9 @@ void ExternalResourcesPage::openedImpl()
     ui->actionsToolbar->setVisibilityState(m_wide_bar_setting->get().toByteArray());
+
+    // Enable the symbolic link warning when the folder is a symbolic link
+    ui->isSymlinkWarning->setVisible(FS::isSymLink(m_model->dir().absolutePath()));
```
same as for worlds

Trial97 commented on launcher/ui/pages/instance/InstanceSettingsPage.ui:
```diff
@@ -583,7 +583,7 @@
         <spacer name="verticalSpacer_2">
          <property name="orientation">
-          <enum>Qt::Vertical</enum>
+          <enum>Qt::Orientation::Vertical</enum>
```
Can you revert all the lines that add the extra specification in UI files?
This looks like is causing some weird alignments on the pages(that are not present on develop)

Trial97 commented on launcher/minecraft/UpdateGlobalDirectoriesTask.h:
```diff
@@ -0,0 +1,22 @@
+class UpdateGlobalDirectoriesTask : public Task {
+   public:
+    explicit UpdateGlobalDirectoriesTask(MinecraftInstance* inst, QWidget* parent = 0);
+    virtual ~UpdateGlobalDirectoriesTask();
```
```suggestion
    virtual ~UpdateGlobalDirectoriesTask() = default;
```

Trial97 commented on launcher/FileSystem.cpp:
```diff
@@ -285,6 +285,16 @@ bool ensureFolderPathExists(const QString folderPathName)
+bool checkFolderPathExists(const QString& folderPathName)
+{
+    return QDir(folderPathName).exists();
+}
```
I'm not sure we need this(just use QFileInfo::exists)

Trial97 commented on launcher/minecraft/MinecraftInstance.cpp:
```diff
@@ -1229,4 +1235,11 @@ QList<Mod*> MinecraftInstance::getJarMods() const
+{
+    // Global directories
+    m_update_global_directories_task = std::make_shared<UpdateGlobalDirectoriesTask>(this, nullptr);
+    m_update_global_directories_task->start();
```
You do not use the result of the task so do not make it a task

Trial97 commented on launcher/minecraft/UpdateGlobalDirectoriesTask.h:
```diff
@@ -0,0 +1,22 @@
+
+   private:
+    MinecraftInstance* m_inst;
+    QWidget* m_parent;
```
This is passed as nullptr remove it

Trial97 commented on launcher/minecraft/UpdateGlobalDirectoriesTask.cpp:
```diff
@@ -0,0 +1,187 @@
+    QString m_destination;
+    MinecraftInstance* m_inst;
+    QString m_setting;
+    QWidget* m_parent;
```
You should not mix tasks with the UI.
Use signals or pass helper structures similar to ATlauncher instance creation

Trial97 commented on launcher/minecraft/UpdateGlobalDirectoriesTask.cpp:
```diff
@@ -0,0 +1,187 @@
+        FileConflictDialog dialog(source, destination, true, parent);
+        FileConflictDialog::Result result = dialog.execWithResult();
+
+        if (result == FileConflictDialog::Cancel)
+            return false;
+        else if (result == FileConflictDialog::ChooseDestination)
+            return FS::deletePath(source);
+        else if (result == FileConflictDialog::ChooseSource)
+            FS::deletePath(destination);
```
change this to switch

Trial97 reviewed (commented):
not tested it yet but the tasks spaghetti caught my eye.
Basically, I think the steps in UpdateGlobalDirectoriesTask can be done more easily with QtConcurent if you do not care about the result(like in your case you run the task but do not display the task result)
Also in case you have multiple screenshots folders already with data(can be any of the 3) and you globally enable the setting you are bombarded with a dialog for each instance and that is not fun.
You should find a way to ask user once and apply for all

theminecoder commented:
> > I noticed that you marked it as being ready so I did a bit of code review. havn't tested yet.
> > You using a per instance setting (which would be desired behavior), but your not providing a global default setting.
> > I think that this should be changed so that there is a global default settings for the folders with an instance setting that overrides the global defaults to use the instance's settings. Look at the java override setting for an example of what I mean.
> 
> Yeah I was planning on adding that eventually, but I haven't exactly figured out how I want that to look. As in, if you enable the default checkbox for the global saves folder, and you have a lot of instances with "New World", you might get a lot of FileConflictDialogs thrown at you. If you think that's not a problem, I'll go ahead and implement it, else I've got to come up with a good solution.

Love the concept here, would be a huge time saver. Thinking of ideas of how you could handle the initial migration of items: If you were to come across a naming conflict (or you could just provide this as an option in general - it may help with organisation) you could look at prefixing the resulting file with the original instance name as such "Instance_A_Screenshotxxxxx.png"

Naomi-alt committed:
Merge branch 'develop' into global-folders

Naomi-alt committed:
Change global folders to shared folders

- For each instance, a custom (shared) screenshots/worlds/resourcepacks folder can be set
- Remove ability to set shared folders for all instances (Might add later in different form)

Signed-off-by: Naomi <103967@gmail.com>

Naomi-alt committed:
Review feedback

Signed-off-by: Naomi <103967@gmail.com>

Naomi-alt committed:
Restructure updating of shared directories

Signed-off-by: Naomi <103967@gmail.com>

Naomi-alt committed:
Clang-format

Signed-off-by: Naomi <103967@gmail.com>

Naomi-alt commented:
I've found some time to work on it again! :D

I decided to pivot a bit and simplify it a bit by only adding shared folder to instances, and not adding global shared folders.
I might do that later, but as a first step this is probably better.

I've moved some functions to `FileSystem.h`, but I think they don't belong there since they use a UI? Not sure where else to put them though since they are generic and thus also don't really belong in `MinecraftInstance.cpp`

Trial97 commented on launcher/FileSystem.cpp:
```diff
@@ -1695,4 +1740,76 @@ QString getUniqueResourceName(const QString& filePath)
+            return true;
+        }
+
+        FS::deletePath(destination);
```
Discussed this on Discord a few days ago:
- my opinion is that this should at least not affect old instances(when you enable the setting, the content of the folder on the old instance should not be overridden by the symlink); maybe just make it somehow relevant to the the new instances 
- @getchoo did not like the idea of sharing folders between different instances and suggested just using the copy functionality and then overriding the stuff not needed https://discord.com/channels/1031648380885147709/1031823065937629267/1381960047382368379

Naomi-alt replied:
1. Not sure if I follow completely? The setting is now per-instance only, so you can only change 1 instance at a time. When you enable it, existing files are copied to the shared folder, including conflict checks. I could change it so only empty/non-existing folders can be changed to a shared folder?
2. So by copying you mean something like copying all files from the shared folder to the instance folder, and then when the instance is stopped copy everything back to the shared folder? Unless I'm misunderstanding this, it would mean you would still have multiple copies of potentially a lot of files of significant size, almost defeating the whole purpose of having a shared folder in the first place?

Trial97 replied:
1. Yes, asking the user what to do in case of a non-empty folder would be great.
2. PrismLauncher has a Copy dialog that allows symlinks, and the suggestion is just to use that functionality to create new instances with shared resources. Or even let the user manually create the symlinks outside Prism.

Ryex replied:
On (2) I'm not finding discussion in question on discord but I'm not sure it's the solution you imply getchoo thinks it is. the ability to link resource folders without needing to set up a template instance is much more user friendly and strait forward.

Trial97 replied:
It was somewhere around here: https://discord.com/channels/1031648380885147709/1031823065937629267/1381942517536264265
(I started as I wanted to implement the options.txt handling, then derived to the instance template, then part of the templates, I mentioned this PR as part of symlinking the folders)
https://discord.com/channels/1031648380885147709/1031823065937629267/1381961967979204609

Naomi-alt replied:
> Or even let the user manually create the symlinks outside Prism.

Then I might as well throw this whole feature out of the door, since the whole point of it is to give non-technical users the ability to easily create symlinks without having to know what they are 🙃

Ryex replied:
> > Or even let the user manually create the symlinks outside Prism.
> 
> Then I might as well throw this whole feature out of the door, since the whole point of it is to give non-technical users the ability to easily create symlinks without having to know what they are 🙃

This. 

sure there is technically overlap with the copy dialog. but _only_ if the target shared path is *inside* another instance (come to think of it  that might be worth detecting and handling - add the appropriate linked instance tags so deletions will warn the user.).

If the target isn't inside another instance it allows the creation of shared shader and resource pack folders in a easy to understand way. I'm all for that kind of feature.

Trial97 replied:
I'm fine with this being a setting, but it should allow a revert back when the user unchecks the setting(either the link is removed, or a backup folder replaces it). If that is not desired, I would prefer it would just be an action a user can do, not a setting; the user clicks something, and that lets them link the folders.

Also, as Ryex mentioned, there is a copy dialog that allows the users to symlink folders in instances:
It would be nice in that case, the shared folder setting to be populated and in case user adds/changes shared resources to update the settings that dialog touches so when removing the instance that has the original shared folder(not the link) the user should be notified about the instances that may break.

Naomi-alt replied:
When the user unchecks the setting, the link is indeed removed.

Populating the shared folder settings when copying using symlinks is indeed a good idea, I'll look into that.

ninetailedtori replied:
Population of shared folder with symlinks is a good idea, do we aim to continue the edit on this then? Is development on this PR still occurring or is it better to have someone adopt it? Just checking :]

Naomi-alt replied:
I was waiting for feedback that never came, so it's not completely abandoned by me.
(Unless I remembered/interpretted the situation wrong)

AlphaLeoli replied:
I've been tracking this thread for a very long time. I'm glad to see some activity on it again :sob: 
I've been using a python script as my main launcher, but if this were a feature I'd switch immediately

ninetailedtori replied:
w.r.t. the main issue of this question, I think the best option would actually be to prompt the user with a query listing all instances that exist that HAVE images already in them, and ask if they're OK with us retroactively generating the symlink(/hardlink/move all screenshots to the unified folder after instance closes) for these automatically. In that case, we can provide a separate function as well, that MOVES over all the screenshots to the destination unified folder (whether categorised by instance name/version, which would be good), then wipes the instance folder and generates the symlink to the destination one. It's not too out there to do that, I think. Existing folders that are empty should be automatically overwritten, and for all future instances handle as policy dictates.

Worth noting: if you want to have a per-instance policy whether or not to use symlink/hardlink/etc, you end up in that awkward place of having to track instance-specific flags, which slows time down. Having a system-wide screenshot policy then providing a means to enforce it safely would be ideal imo. But the whole "shared folder" system is opt in anyway, and people who'd opt into it wouldn't really make sense of not doing it for some instances, imo. It's best to enforce it system-wide if they opt into the shared folder policy.

ninetailedtori replied:
I haven't had the fullest of time to check the code, but I'll check it in a bit, just working on my patch for gamescope rn :]

OffsetMonkey538 commented:
How granular exactly is the configuration of the folder structure here currently? I'd love to be able to send screenshots to `$XDG_PICTURES_DIR/Screenshots/Minecraft/$INSTANCE_NAME` for example

Trial97 commented:
@Naomi-alt is this ready for review?
If yes can you please resolve the conflicts?

Trial97 commented:
closing this for now.
If you or somebody else wants to pick this up feel free to open a new Pull Request.
We can reopen this PR but just let me know that you are still working on this.

pers-ephone commented:
Does anyone know how this does/did plan to tackle the problem of windows symlink needing additional permissions (see https://github.com/PrismLauncher/PrismLauncher/issues/38#issuecomment-1399183603 ) ?
I am interested in this PR or other solutions but this point seems like a significant problem to me.

Linx-ESP commented:
> Does anyone know how this does/did plan to tackle the problem of windows symlink needing additional permissions (see [#38 (comment)](https://github.com/PrismLauncher/PrismLauncher/issues/38#issuecomment-1399183603) ) ? I am interested in this PR or other solutions but this point seems like a significant problem to me.

Does it? I can run `New-Item -Type SymbolicLink -Target .\Downloads\aaaa -Path .\Downloads\cccc` without admin privileges.  
Maybe hard link or junctions are the ones that do require
